feat: cicd actions

s4005236 opened the pull request:
This PR includes changes in order to implement GitHub Actions and local pre-commit hooks correctly.

---

To get these functions to work, a restructure of the project was necessary. This however shouldn't change the preexisting workflow too much.

pre-commit hooks:
- black (code formatter)

GitHub Actions:
- pylint (checks code in comparison to the PEP 8 standard, fails if codebase has lower quality than 8.0/10 points)

Note: To get the functions to work, some testing was necessary. This also included committing and pushing to the repo. That's why the number of commits is that high. :)

Duspli commented on README.md:
Could you change the Open Gangster to WAVE please :)

Duspli commented on README.md:
```diff
@@ -1 +1,10 @@
 # Open Gangster
```
Could you change Open Gangster to WAVE pls? :)

FGBxRamel commented on .pylintrc:
```diff
@@ -119,15 +119,15 @@ unsafe-load-any-extension=no
 
 # Naming style matching correct argument names.
-argument-naming-style=camelCase
+argument-naming-style=snake_case
```
And to add the @Duspli s comment: Is this documented somewhere? Using snake case adheres to PEP 8, but is this documented?

FGBxRamel replied:
Resolved by [docs: add code styling](https://github.com/s4005236/wave/pull/16/commits/e67f03653fdcb50e3044bc06ca8c87482ab41fb0)

FGBxRamel commented on pyproject.toml:
```diff
@@ -1,33 +1,20 @@
-    { path = "device"},
-    { path = "image"},
-    { path = "ui"}
+    "Raziel Kretschmer <s4005251@ba-sachsen.de>",
+    "Alexander Jahn <s4005236@ba-sachsen.de>",
+    "Maximilian Unger <s4005395@ba-sachsen.de>",
+    "Dustin Kowarsch <s4005362@ba-sachsen.de>"
```
Please change the email addresses to the new domain format

FGBxRamel commented on src/core/src/core.py:
What exactly are these changes? Why are they necessary?

s4005236 replied:
The changes to some python files (e.g. core.py, gestures.py) are done automatically by the black formatter. They don't add or change anything when it comes to the functionalities of the files. Only the imports are corrected because of the structure change in order for them to work as intended.

If these style changes should not be in this commit, it would also be possible to revert them. They will get formatted again once someone commits changes to these files.

FGBxRamel reviewed (changes requested):
Good work on the config files. 

*Please merge less commits!*. Please use [tools for rewriting git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) and [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force) to cleanup the commits.

s4005236 commented:
> Good work on the config files.
> 
> _Please merge less commits!_. Please use [tools for rewriting git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) and [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force) to cleanup the commits.

In your opinion @FGBxRamel , what is good number of commits to put in a merge? (asking for this and future PRs)

FGBxRamel reviewed (commented):
Changes lgtm; still open question: Why do we have 4 setups files?

FGBxRamel commented:
> > Good work on the config files.
> > _Please merge less commits!_. Please use [tools for rewriting git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) and [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force) to cleanup the commits.
> 
> In your opinion @FGBxRamel , what is good number of commits to put in a merge? (asking for this and future PRs)

@s4005236 Sorry, my phrasing was not quite accurate. It's not really about some made up number, but more about a clean history. For example:
The commits: [test: let pylint be correct](https://github.com/s4005236/wave/pull/16/commits/ef3a6b34917c2d6a78dd775305e0031f9c3a1ae8) and [rev: rev config from test](https://github.com/s4005236/wave/pull/16/commits/df0dce75a4367a1f2c317debb11a62d0d0dead16) should be merged to e.g. "feat: Working pylint config"

"test" is generally only used if unit-tests or something similar is changed or added. Everything else where you *literally* test a config should afterwards be merged / rewritten to make a meaningful commit. This is especially true for when the test is reverted afterwards, as these two actions have no meaning or value standalone in the git history. Note that not all reverted commits have to be merged. Sometimes it's good to leave a revert-commit in the history, just to show others that some approach didn't work.

Another example of the above (but not limited to) would be these three commits:
1. [test: black output](https://github.com/s4005236/wave/pull/16/commits/f24d3995ed69436183b427023fa4f0f7e2d69bff)
2. [test: check black output](https://github.com/s4005236/wave/pull/16/commits/ab2df397b3286848649e62b4e702970ec90e25a7)
3. [style: auto-format project with black](https://github.com/s4005236/wave/pull/16/commits/0959a4dfae182a59cca3542936a9375c3a466c82)

I would argue that it would be good practice to merge these three into the last one, as the two tests are not adding any value.

FGBxRamel reviewed (commented):
Lgtm

s4005236 commented on src/core/setup.py:
The global setup_dev.py is for working with the whole project. When working on a specific submodule, a developer might only open up the corresponding folder. There, they can just run the normal `python setup_dev.py` command like they would at root, without the need to first build the project globally. This would mean opening up project at root, running the command, changing folder back to submodule, which would remove the purpose of the setup-file to make setting up easier.

tl;dr: Make working with the submodules-folders more convenient, remove the need to use the global setup file.

s4005236 commented on src/core/src/core.py:
```diff
@@ -1,13 +1,13 @@
-from ..enums.gesture_types import GestureTypes
-from ..enums.bodyparts import Fingers
 import logging
+from src.core.src.gesture import Gesture
```
The src-structure is necessary for poetry to recognize the modules as packages. Without it, errors occure, e.g. when there is more than one python file at submodule-root. (E.g. currently at main, building the "core" package fails, because it detects "core.py" and "gestures.py".)

Without the structure change, the whole setup wouldn't work as intended because of said error. For the intention of the PR to introduce a working CI/CD pipeline, the structure change was necessary.

s4005236 commented:
> Changes lgtm; still open question: Why do we have 4 setups files? @FGBxRamel 

answered at https://github.com/s4005236/wave/pull/16#discussion_r2084278156

s4005236 commented:
> > > Good work on the config files.
> > > _Please merge less commits!_. Please use [tools for rewriting git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) and [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force) to cleanup the commits.
> > 
> > 
> > In your opinion @FGBxRamel , what is good number of commits to put in a merge? (asking for this and future PRs)
> 
> @s4005236 Sorry, my phrasing was not quite accurate. It's not really about some made up number, but more about a clean history. For example: The commits: [test: let pylint be correct](https://github.com/s4005236/wave/pull/16/commits/ef3a6b34917c2d6a78dd775305e0031f9c3a1ae8) and [rev: rev config from test](https://github.com/s4005236/wave/pull/16/commits/df0dce75a4367a1f2c317debb11a62d0d0dead16) should be merged to e.g. "feat: Working pylint config"
> 
> "test" is generally only used if unit-tests or something similar is changed or added. Everything else where you _literally_ test a config should afterwards be merged / rewritten to make a meaningful commit. This is especially true for when the test is reverted afterwards, as these two actions have no meaning or value standalone in the git history. Note that not all reverted commits have to be merged. Sometimes it's good to leave a revert-commit in the history, just to show others that some approach didn't work.
> 
> Another example of the above (but not limited to) would be these three commits:
> 
> 1. [test: black output](https://github.com/s4005236/wave/pull/16/commits/f24d3995ed69436183b427023fa4f0f7e2d69bff)
> 2. [test: check black output](https://github.com/s4005236/wave/pull/16/commits/ab2df397b3286848649e62b4e702970ec90e25a7)
> 3. [style: auto-format project with black](https://github.com/s4005236/wave/pull/16/commits/0959a4dfae182a59cca3542936a9375c3a466c82)
> 
> I would argue that it would be good practice to merge these three into the last one, as the two tests are not adding any value.
@FGBxRamel 

Rebased the history, squashed all test commits that don't add significant value.

s4005236 commented on docs/DevOps/CODE_STYLING.md:
```diff
@@ -0,0 +1,14 @@
+
+### Locally (pre-commit)
+- **Black**: A code formatter for Python that enforces PEP8 by automatically reformatting code. If formatting issues are found, Black modifies the affected files. The commit fails, and you must stage the changes and retry the commit.
+- **(TODO) isort**: //
```
isort will be implemented in a future PR. This is just a note to not forget about documenting it.

FGBxRamel reviewed (approved):
Lgtm, nice work
